IBAutomater updates

StefanoRaggi opened the pull request:
- Changed `Start()` method from void to bool
- Added new `IsRunning()` method
- Improved error handling
- When an existing session is detected, IBAutomater will not attempt to reconnect anymore
- Added new `Restart()` method
- Updated `Start()` to wait for initialization and to include error handling

StefanoRaggi committed:
IBAutomater updates

- Changed Start() method from void to bool
- Added new IsRunning() method
- Improved error handling

StefanoRaggi committed:
Add checks in Stop() and IsRunning()

StefanoRaggi committed:
Remove logging from IsRunning()

StefanoRaggi committed:
Update IBAutomater C# demo app

StefanoRaggi committed:
Do not attempt to reconnect session if existing session detected

Martin-Molinero commented on CSharpDemoIBAutomater/Program.cs:
```diff
@@ -25,16 +25,32 @@ public static void Main(string[] args)
 
             // Wait a few seconds for startup
             Console.WriteLine("IBAutomater started, waiting 30 seconds");
             Thread.Sleep(30000);
```
nit - maybe the thread sleep could be in an `else` statement

Martin-Molinero commented on QuantConnect.IBAutomater/IBAutomater.cs:
```diff
@@ -101,72 +107,104 @@ public IBAutomater(string ibDirectory, string ibVersion, string userName, string
+            if (_isStarting)
+            {
+                OutputDataReceived?.Invoke(this, new OutputDataReceivedEventArgs("IBAutomater is already starting."));
+                return false;
```
Wonder if this should be `true`? The user wants to start the `automater`, if he is already starting, seems like a happy path 🤔

Martin-Molinero commented on QuantConnect.IBAutomater/IBAutomater.cs:
```diff
@@ -207,6 +250,32 @@ public void Stop()
+
+            try
+            {
+                Process.GetProcessById(_processId);
```
Could also keep the `Process` instance and check `Process.HasExited`?
just checking: the `IBAutomater.bat/IBAutomater.sh` **don't** start another process and return right away, but instead, wait until it finishes?

StefanoRaggi replied:
Correct, the bat/sh process does not terminate until the child process is terminated.

Martin-Molinero commented on QuantConnect.IBAutomater/IBAutomater.cs:
```diff
@@ -101,72 +107,104 @@ public IBAutomater(string ibDirectory, string ibVersion, string userName, string
+                OutputDataReceived?.Invoke(this, new OutputDataReceivedEventArgs($"IBAutomater process started - Id:{process.Id}"));
 
-            process.Start();
+                _processId = process.Id;
```
Maybe should we set `_processId` to 0 when we set starting to `true` so that we avoid any race condition

Martin-Molinero commented on QuantConnect.IBAutomater/IBAutomater.cs:
```diff
@@ -207,6 +250,32 @@ public void Stop()
+        /// <returns>true if the IBGateway is running</returns>
+        public bool IsRunning()
+        {
+            if (_processId == 0)
```
Should we also add a `_isStarting` check here?

Martin-Molinero commented on QuantConnect.IBAutomater/IBAutomater.cs:
```diff
@@ -175,6 +213,11 @@ public void Start(bool waitForExit)
         /// <remarks>The IB Gateway application will be terminated</remarks>
         public void Stop()
         {
+            if (!IsRunning())
```
Wondering if `Start/Stop/IsRunning` should share a lock to avoid thread race issues, since their state/result depends on each other

StefanoRaggi committed:
Address review

- Updated IsRunning to use process.HasExited instead of GetProcessById
- Fixed isStarting checks
- Added thread locking around Start/Stop/IsRunning

StefanoRaggi committed:
Add new Restart method

StefanoRaggi committed:
Remove obsolete isStarting flag

StefanoRaggi committed:
Updated Start() to wait for initialization and to include error handling

Martin-Molinero commented on QuantConnect.IBAutomater/IBAutomater.cs:
```diff
@@ -97,115 +133,225 @@ public IBAutomater(string ibDirectory, string ibVersion, string userName, string
                     }
                 }
+
+                _process = null;
```
Should we try to kill `_process` too?

StefanoRaggi replied:
`_process` is the parent process (sh/bat script) for the processes killed in the previous lines of code and should be terminated automatically.
I did a quick run with `_process?.Kill();` added and this was the error:
```
System.InvalidOperationException: Cannot process request because the process (14300) has exited.
```

Martin-Molinero commented:
Nice work 👍
Pending manual tests regarding the behavior when a competing session is started. It seems this should be detected by the Automater, update `IsRunning()` and move into an unrecoverable invalid state that will fail to start/restart.
Also, I think `_ibAutomaterInitializeEvent` should be cleared before start, just in case

StefanoRaggi committed:
Reset _ibAutomaterInitializeEvent upon Start

StefanoRaggi committed:
Add new GetLastStartResult() method

StefanoRaggi committed:
Make IsRunning() private + add _lastStartResult check in Start()
